File: subprojects/core/src/main/java/hu/bme/mit/theta/core/type/bvtype/BvConcatExpr.java
package hu.bme.mit.theta.core.type.bvtype;

import com.google.common.collect.ImmutableList;
import hu.bme.mit.theta.common.Utils;
import hu.bme.mit.theta.core.model.Valuation;
import hu.bme.mit.theta.core.type.Expr;
import hu.bme.mit.theta.core.utils.TypeUtils;

import java.util.List;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.collect.ImmutableList.toImmutableList;

public final class BvConcatExpr implements Expr<BvType> {
    private static final int HASH_SEED = 8264;
    private static final String OPERATOR_LABEL = "++";

    private final List<Expr<BvType>> ops;

    private volatile int hashCode = 0;

    private BvConcatExpr(final Iterable<? extends Expr<BvType>> ops) {
        checkNotNull(ops);
        checkArgument(ops.iterator().hasNext());
        this.ops = ImmutableList.copyOf(ops);
    }

    public static BvConcatExpr of(final Iterable<? extends Expr<BvType>> ops) {
        return new BvConcatExpr(ops);
    }

    public static BvConcatExpr create(final List<? extends Expr<?>> ops) {
        checkNotNull(ops);
        return BvConcatExpr.of(ops.stream().map(TypeUtils::castBv).collect(toImmutableList()));
    }

    @Override
    public int getArity() {
        return ops.size();
    }

    @Override
    public BvType getType() {
        return BvType.of(ops.stream().map(o -> o.getType().getSize()).reduce(0, Integer::sum));
    }

    @Override
    public BvLitExpr eval(final Valuation val) {
        return getOps().stream().skip(1).reduce(
            (BvLitExpr) getOps().get(0).eval(val),
            (op1, op2) -> (op1.concat((BvLitExpr) op2.eval(val))),
            BvLitExpr::concat
        );
    }

    @Override
    public List<? extends Expr<BvType>> getOps() {
        return ops;
    }

    @Override
    public Expr<BvType> withOps(List<? extends Expr<?>> ops) {
        checkNotNull(ops);
        checkArgument(ops.size() >= 1);

        return of(ops.stream().map(TypeUtils::castBv).collect(toImmutableList()));
    }

    @Override
    public int hashCode() {
        int result = hashCode;
        if (result == 0) {
            result = HASH_SEED;
            result = 31 * result + getOps().hashCode();
            hashCode = result;
        }
        return result;
    }

    @Override
    public boolean equals(final Object obj) {
        if (this == obj) {
            return true;
        } else if (obj instanceof BvConcatExpr) {
            final BvConcatExpr that = (BvConcatExpr) obj;
            return this.getOps().equals(that.getOps());
        } else {
            return false;
        }
    }

    @Override
    public String toString() {
        return Utils.lispStringBuilder(OPERATOR_LABEL).addAll(getOps()).toString();
    }
}

File: subprojects/common/core/src/main/kotlin/hu/bme/mit/theta/core/type/bvtype/BvConcatExpr.kt
/*
 *  Copyright 2025 Budapest University of Technology and Economics
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package hu.bme.mit.theta.core.type.bvtype

import hu.bme.mit.theta.common.Utils
import hu.bme.mit.theta.core.model.Valuation
import hu.bme.mit.theta.core.type.Expr
import hu.bme.mit.theta.core.utils.TypeUtils
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
@SerialName("BvConcat")
data class BvConcatExpr(override val ops: List<Expr<BvType>>) : Expr<BvType> {

  init {
    check(ops.isNotEmpty())
  }

  companion object {

    private const val OPERATOR_LABEL = "++"

    @JvmStatic fun of(ops: Iterable<Expr<BvType>>) = BvConcatExpr(ops.toList())

    @JvmStatic fun create(ops: List<Expr<*>>) = BvConcatExpr(ops.map { TypeUtils.castBv(it) })
  }

  override val type: BvType = BvType.of(ops.sumOf { it.type.size })

  override fun eval(`val`: Valuation): BvLitExpr =
    ops.drop(1).fold(ops[0].eval(`val`) as BvLitExpr) { acc, op ->
      acc.concat(op.eval(`val`) as BvLitExpr)
    }

  override fun withOps(ops: List<Expr<*>>): Expr<BvType> = of(ops.map { TypeUtils.castBv(it) })

  override fun toString(): String =
    Utils.lispStringBuilder(OPERATOR_LABEL).body().addAll(ops).toString()
}
